rs-matter/src/transport/core.rs

```diff
@@ -103,680 +222,890 @@ impl<'a> Matter<'a> {
+            let mut accept_timeout = pin!(self.with_locked(&self.rx, |packet| {
+                self.handle_accept_timeout_rx_packet(packet).then_some(())
+            }));
 
-        if let Either::First(result) = result {
-            if let Err(e) = &result {
-                error!("Exitting RX loop due to an error: {:?}", e);
-            }
+            let mut timer = pin!(Timer::after(embassy_time::Duration::from_millis(50)));
 
-            result?;
+            select(&mut accept_timeout, &mut timer).await;
         }
-
-        Ok(())
     }
 
-    #[inline(always)]
-    pub async fn handle_tx<S>(&self, mut send: S) -> Result<(), Error>
-    where
-        S: NetworkSend,
-    {
+    async fn process_orphaned_rx(&self) -> Result<(), Error> {
         loop {
-            loop {
-                {
-                    let mut send_buf = self.tx_buf.get().await;
+            info!("Waiting for orphaned RX packets");
 
-                    let mut tx = alloc!(Packet::new_tx(&mut send_buf));
+            self.with_locked(&self.rx, |packet| {
+                self.handle_orphaned_rx_packet(packet).then_some(())
+            })
+            .await;
+        }
+    }
 
-                    if self.pull_tx(&mut tx)? {
-                        let addr = tx.peer;
+    async fn process_dropped_exchanges(&self) -> Result<(), Error> {
+        loop {
+            trace!("Waiting for dropped exchanges");
 
-                        let start = tx.get_writebuf()?.get_start();
-                        let end = tx.get_writebuf()?.get_tail();
+            let mut tx = self.get_if(&self.tx, |packet| packet.buf.is_empty()).await;
+            tx.clear_on_drop(true); // In case of error, or if the future is dropped
 
-                        send.send_to(&send_buf[start..end], addr).await?;
-                    } else {
-                        break;
-                    }
+            let wait = match self.handle_dropped_exchange(&mut tx) {
+                Ok(wait) => {
+                    tx.clear_on_drop(false);
+                    wait
                 }
-            }
+                Err(e) => {
+                    error!("UNEXPECTED RX ERROR: {e:?}");
+                    false
+                }
+            };
 
-            self.wait_tx().await?;
+            drop(tx);
+
+            if wait {
+                let mut timeout = pin!(Timer::after(embassy_time::Duration::from_millis(100)));
+                let mut wait = pin!(self.dropped.wait());
+
+                select(&mut timeout, &mut wait).await;
+            }
         }
     }
 
-    #[inline(always)]
-    pub async fn handle_rx_multiplex<'t, 'e, const N: usize, R>(
-        &'t self,
-        mut receiver: R,
-        sts_buf: &mut [u8; MAX_RX_STATUS_BUF_SIZE],
-        construction_notification: &'e Notification,
-        channel: &Channel<NoopRawMutex, ExchangeCtr<'e>, N>,
-    ) -> Result<(), Error>
+    async fn handle_rx_packet<const N: usize, S>(
```

Comment:
this needs some documentation. In particular I find a `handle() -> bool` as a concept returning "handled" or "not handled" however it is not immediately obvious about the difference between `not handled` and `error` (since error means not handled due to an error).

Could this return some named enum instead where we could say results like `Handled` or `Duplicate` or other potential values?

Reply:
I would not do this for now. This is an internal method (non-API). Also, the caller of this method does not need to actually know _why_ the packet was not handled (i.e. "duplicate" or other reason). It just needs to know that the packet was not handled so as to drop the packet and that's it. Or if an error had occurred, to log the error and drop the packet.